Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,505 @@
+        var digits: [UInt8] = []
+
+        if decimalString.utf8.count > 1 {
+            for digit in decimalString {
+                if digit == Character(".") {
+                    continue
+                }
+                guard (Character("0")...Character("9")).contains(digit) else {
+                    throw BSONError.InvalidArgumentError(
+                        message: "Syntax Error: \(digit) is not a digit '0'-'9'"
+                    )
+                }
+                if digit == Character("0") && leadingZero {
+                    continue
+                }
+                if digit != Character("0") && leadingZero {
+                    // seen a non zero digit
+                    leadingZero = false
+                }
+                guard let digitValue = UInt8(String(digit)) else {
+                    throw BSONError.InvalidArgumentError(
+                        message: "Syntax Error: \(digit) cannot be represented in a UInt8"
+                    )
+                }
+                digits.append(digitValue)
+            }
+        } else if decimalString.utf8.count == 1 {
+            guard let digitValue = UInt8(String(decimalString[decimalString.startIndex])) else {
+                throw BSONError.InvalidArgumentError(
+                    message: "Syntax Error: \(decimalString[decimalString.startIndex]) cannot be represented in a UInt8"
+                )
+            }
+            digits.append(digitValue)
+        }
+        return digits
+    }
+
+    private func toString() -> String {
+        // swiftlint:disable:previous cyclomatic_complexity
+        var exponent: Int
+        var significandPrefix: Int
+
+        // If the combination field starts with 0b11 it could be special (NaN/Inf)
+        if self.value.hi.getBits(1...2) == 0b11 {
+            if self.isInfinity {
+                return (self.isNegative ? "-" : "") + "Infinity"
+            }
+            if self.isNaN {
+                return "NaN"
+            }
+            // The number is neither NaN nor Inf
+            // Decimal interchange floating-point formats c,2,ii
+            exponent = Int(self.value.hi.getBits(3...(Self.exponentLength + 2)))
+            significandPrefix = Int(self.value.hi.getBit(20) + 0b1000)
+        } else {
+            // Decimal interchange floating-point formats c,2,i
+            exponent = Int(self.value.hi.getBits(1...Self.exponentLength))
+            significandPrefix = Int(
+                self.value.hi.getBits(
+                    (Self.exponentLength + 1)...(Self.exponentLength + Self.combinationLength)
+                )
+            )
+        }
+
+        exponent -= Self.exponentBias
+
+        var significand128 = UInt128()
+
+        // significand prefix (implied bits) combined with removing the combination and sign fields
+        significand128.hi = UInt64((significandPrefix & 0xF) << 46) | self.value.hi.getBits(18...63)
```

Comment:
Sinnce the higher order bits of `value.hi` will be 0, we don't need to do `getBits(18...63)` here, right?

Reply:
The high order bits can be nonzero here, the sign bit along with the exponent field is present.

Reply:
Oh right, of course.